### PR TITLE
Merges the Trace and Logging module with autoconfigure

### PR DESCRIPTION
**Pull request by joaoandremartins**
- Adds Logging README under the reference document in Spring Cloud Sleuth,
- Changes name of `StackdriverJsonLayoutLoggerTests` so it's not affected by changes to the `org.springframework.cloud.gcp` logger,
- Changes `StackdriverTraceReporterTests` to use `ApplicationContextRunner` for tests

### CODE REVIEW

File: spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/trace/sleuth/StackdriverTraceSpanListenerTests.java [hunk 1]
```diff
@@ -47,8 +47,9 @@
 import org.springframework.mock.env.MockEnvironment;
 import org.springframework.test.context.junit4.SpringRunner;
 
-@SpringBootTest(classes = {
-		StackdriverTraceSpanListenerTests.TestConfiguration.class })
+@SpringBootTest(
```
Comment (joaoandremartins) [inline spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/trace/sleuth/StackdriverTraceSpanListenerTests.java]: @artembilan @snicoll @dsyer these tests are activating `GcpCloudSqlAutoConfiguration` for some reason, even though it's not listed in `classes`.
The other "sibling" autoconfig test, `StackdriverTraceAutoConfigurationTests` isn't enabling `GcpCloudSqlAutoConfiguration`.
I can't figure out what the difference is. Would you know off the top of your head?

Comment (artembilan) [reply spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/trace/sleuth/StackdriverTraceSpanListenerTests.java]: @joaoandremartins ,

look, in the end of this class you have a configuration class:
```
@Configuration
@EnableAutoConfiguration
static class TestConfiguration {
```
So, that `@EnableAutoConfiguration` definitely triggers all the possible auto-configurations to be loaded.

@snicoll , suggested once to follow the Spring Boot way to test your requirements. Like load only those configurations you need for this particular test. For example I did this in one of my tests: https://github.com/spring-projects/spring-integration-java-dsl/blob/master/src/test/java/org/springframework/integration/dsl/test/jpa/JpaTests.java#L170

**Comment by meltsufin** [comment]
Should we be really merging Trace and Logging into auto-configure?

File: spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/trace/sleuth/StackdriverTraceSpanListenerTests.java (hunk 1 above)
Comment (joaoandremartins) [reply spring-cloud-gcp-autoconfigure/src/test/java/org/springframework/cloud/gcp/autoconfigure/trace/sleuth/StackdriverTraceSpanListenerTests.java]: Thanks a lot for the tip @artembilan !

> @snicoll , suggested once to follow the Spring Boot way to test your requirements. Like load only those configurations you need for this particular test. For example I did this in one of my tests: https://github.com/spring-projects/spring-integration-java-dsl/blob/master/src/test/java/org/springframework/integration/dsl/test/jpa/JpaTests.java#L170

That's already in scope for #313 . Let's take care of all of them in a single stroke.

**Comment by joaoandremartins** [comment]
After this change, in order to use the Stackdriver Trace integration, a user will have to have this in their `pom.xml`:

```
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-gcp-autoconfigure</artifactId>
		</dependency>
		<dependency>
			<groupId>com.google.cloud</groupId>
			<artifactId>google-cloud-trace</artifactId>
		</dependency>
		<dependency>
			<groupId>com.google.cloud.trace.v1</groupId>
			<artifactId>sink</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.cloud</groupId>
			<artifactId>spring-cloud-starter-sleuth</artifactId>
			<version>2.0.0.M5</version>
		</dependency>
```

What do we do for cases like these, where multiple dependencies are required? I don't think it's a very good experience to require these 3 dependencies besides `autoconfigure`.
Have a starter to make lives easier?

**Comment by dsyer** [comment]
> What do we do for cases like these, where multiple dependencies are required?

I don't understand why that is necessary. Is there a sample app where this code is used? Can you explain why that isn't either a) the only choice, or b) a choice that is facilitated by a  starter?

**Comment by joaoandremartins** [comment]
I haven't eliminated the starters for logging or trace yet, but you can go to `spring-cloud-gcp-trace-sample` and [replace `spring-cloud-gcp-starter-trace` by the 4 dependencies I quoted](https://github.com/spring-cloud/spring-cloud-gcp/blob/master/spring-cloud-gcp-samples/spring-cloud-gcp-trace-sample/pom.xml#L25).

I think for cases like trace, a starter that facilitates the choice makes a lot of sense.

**Comment by dsyer** [comment]
That was my option b). I'm not 100% convinced that a) isn't better.

**Comment by joaoandremartins** [comment]
@artembilan @balopat @ChengyuanZhao PTAL
@snicoll @dsyer FYI

**Comment by joaoandremartins** [comment]
Ping!
Would be nice to have a review here. Thanks!

**Comment by artembilan** [comment]
I can't say too much about these features, but was really a question to merge everything into the `autocinfigure`? Maybe idea wsa to merge `logging` and `trace` modules? I don't see any values to have them in the `autoconfigure`...

**Comment by joaoandremartins** [comment]
This goes back to the old issue of having less modules in our project. I believe @snicoll and @dsyer brought it up originally.
Since trace and logging probably aren't going to be used standalone, the idea was to have them merged with autoconfigure.
Unlike Pub/Sub and Spanner, where we provide the template class, and it makes sense to have them as separate modules.

File: docs/src/main/asciidoc/trace.adoc [hunk 2]
```diff
@@ -96,3 +96,162 @@ Spring Cloud GCP Trace does override some Sleuth configurations:
...
+`STACKDRIVER_LOG_NAME` and `STACKDRIVER_LOG_FLUSH_LEVEL` environment variables can be used to customize the
+`STACKDRIVER` appender.
+
+Also see the link:../spring-cloud-gcp-starters/spring-cloud-gcp-starter-logging[spring-cloud-gcp-starter-logging] module.
+
+===== Log via Console
+
+If you run your Spring Boot application in a Kubernetes cluster in the Google Cloud (GKE) or on App Engine flexible
+environment, you can log JSON directly from the console by including
+`org/springframework/cloud/gcp/logging/logback-json-appender.xml`. The traceId will be set
```
[the remark this answers is not in the document]
Comment (joaoandremartins) [reply docs/src/main/asciidoc/trace.adoc]: The package name in the resources actually didn't change.

File: spring-cloud-gcp-logging/pom.xml [hunk 3]
```diff
@@ -1,43 +0,0 @@
...
-            <artifactId>spring-webmvc</artifactId>
-        </dependency>
-        <dependency>
-            <groupId>javax.servlet</groupId>
-            <artifactId>javax.servlet-api</artifactId>
-            <scope>provided</scope>
-        </dependency>
-        <dependency>
-            <groupId>com.google.cloud</groupId>
-            <artifactId>google-cloud-logging-logback</artifactId>
```
[the remark this answers is not in the document]
Comment (joaoandremartins) [reply spring-cloud-gcp-logging/pom.xml]: It's there.

**Comment by joaoandremartins** [comment]
Where do you see a `@ConditionalOnClass` for `GcpTraceProperties`?

Btw, this PR got merged already. Will apply the remaining fixes separately.

File: docs/src/main/asciidoc/trace.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (meltsufin) [reply docs/src/main/asciidoc/trace.adoc]: But shouldn't it, to be consistent with the package name of the classes?

[the remark this answers is not in the document]
Comment (joaoandremartins) [reply docs/src/main/asciidoc/trace.adoc]: I don't feel strongly about it.
There isn't a hard link between the logback file and any java package, but I'll make that change if you require it.

**Comment by meltsufin** [comment]
Sorry, I think I was looking at a stale branch locally. I didn't realize I was reviewing a PR that was already merged.

File: docs/src/main/asciidoc/trace.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (meltsufin) [reply docs/src/main/asciidoc/trace.adoc]: I don't _require_ it. :smile: 
But yeah, I think it makes sense to make it part of the same package as the classes.

**Comment by stevenschlansker** [comment]
I was just looking into using `StackdriverJsonLayout` and I must say I'm confused as to why these artifacts were merged.  I wish I could just depend on `spring-gcp-logging`.  We have a policy of not using autoconfiguration, so now I am in a position where I am considering copy and pasting the logging code into our project, so as not to run into dependency management issues like above.  I think having the logging utilities as a separate artifact is useful.  Auto-configuration is evil :)

**Pull request merged**